src/textual/suggester.py

```diff
@@ -0,0 +1,84 @@
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Iterable, Optional
+
+from .message import Message
+
+if TYPE_CHECKING:
+    from .widgets import Input
+
+
+@dataclass
+class SuggestionReady(Message):
+    """Sent when a completion suggestion is ready."""
+
+    input_value: str
+    """The input value that the suggestion was for."""
+    suggestion: str
+    """The string suggestion."""
+
+
+class Suggester(ABC):
+    """Defines how [inputs][textual.widgets.Input] generate completion suggestions.
+
+    To define a custom suggester, subclass `Suggester` and implement the async method
+    `get_suggestion`.
+    See [`SuggestFromList`][textual.suggester.SuggestFromList] for an example.
+    """
+
+    async def get(self, input: "Input", value: str) -> None:
```

Comment:
Afraid I would veto the name `get`. Too generic. Is this going to be called by the dev? If not, it could be private.

Also we probably don't want to limit ourselves to `Input` widgets. We might want to offer suggestions elsewhere. Since we're posting messages, it could just be a `MessageTarget`